### ISSUE TITLE
staticcheck ignores build-tags

### ISSUE THREAD

**Issue by srebhan** [golangci/golangci-lint#6754]
### Welcome

- [x] Yes, I'm using a binary release within 2 latest [releases](https://github.com/golangci/golangci-lint/releases). Only such installations are supported.
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] Yes, I've read the `typecheck` section of the [FAQ](https://golangci-lint.run/docs/welcome/faq/#why-do-you-have-typecheck-errors).
- [x] Yes, I've tried with the standalone [linter](https://golangci-lint.run/docs/linters/) if available (e.g., gocritic, go vet, etc.).
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

go install

### Description of the problem

When running the static-check linter against platform-dependent code you get false positives like

```console
main.go:14:2: SA4023(related information): the lhs of the comparison is the 2nd return value of this function call (staticcheck)
        msg, err := example(name)
        ^
main.go:15:5: SA4023: this comparison is always true (staticcheck)
        if err != nil {
           ^
2 issues:
* staticcheck: 2
```

if the current platform always returns an error (e.g. to denote the functionality is not supported) but other platforms do not. This is often necessary when implementing OS-near functionality.

The issue seems to be that the linter is _not_ aware of the build-tags and the fact that the code-path may look different on other configurations or platforms. The issue here is that you even cannot silence the linter with a `nolint` directive as this will cause the `nolintlint` linter to fail on other platforms.

### Version of golangci-lint

golangci-lint has version 2.13.1 built with go1.27.0-X:nodwarf5 from (unknown, modified: ?, mod sum: "h1:RuM4OcluM4xFQcGuRE6R7jA33pqxK/W1EsBxpugdZjg=") on (unknown)

### Configuration

<details>
```console
golangci-lint run --enable-only staticcheck  
```
</details>

### Go environment

<details>
```console
go version go1.27.0-X:nodwarf5 linux/amd64
AR='ar'
CC='gcc'
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_ENABLED='1'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
CXX='g++'
GCCGO='gccgo'
GO111MODULE='auto'
GOAMD64='v1'
GOARCH='amd64'
GOAUTH='netrc'
GOBIN='/home/sven/go/bin'
GOCACHE='/home/sven/.cache/go-build'
GOCACHEPROG=''
GODEBUG=''
GOENV='/home/sven/.config/go/env'
GOEXE=''
GOEXPERIMENT='nodwarf5'
GOFIPS140='off'
GOFLAGS=''
GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build16918273=/tmp/go-build -gno-record-gcc-switches'
GOHOSTARCH='amd64'
GOHOSTOS='linux'
GOINSECURE=''
GOMOD='/home/sven/Development/InfluxData/tmp/golangci-lint-bug/go.mod'
GOMODCACHE='/home/sven/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPACKAGESDRIVER=''
GOPATH='/home/sven/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/usr/lib/go'
GOSUMDB='sum.golang.org'
GOTELEMETRY='local'
GOTELEMETRYDIR='/home/sven/.config/go/telemetry'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/usr/lib/go/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.27.0-X:nodwarf5'
GOWORK=''
PKG_CONFIG='pkg-config'
```
</details>

### Verbose output of running

<details>
```console
INFO golangci-lint has version 2.13.1 built with go1.27.0-X:nodwarf5 from (unknown, modified: ?, mod sum: "h1:RuM4OcluM4xFQcGuRE6R7jA33pqxK/W1EsBxpugdZjg=") on (unknown) 
INFO [config_reader] Config search paths: [./ /home/sven/Development/InfluxData/tmp/golangci-lint-bug /home/sven/Development/InfluxData/tmp /home/sven/Development/InfluxData /home/sven/Development /home/sven /home /] 
INFO [config_reader] Module name "github.com/srebhan/golangci-lint-bug" 
INFO [goenv] Read go env for 3.132613ms: map[string]string{"GOCACHE":"/home/sven/.cache/go-build", "GOROOT":"/usr/lib/go"} 
INFO [lintersdb] Active 5 linters: [errcheck govet ineffassign staticcheck unused] 
INFO [loader] Go packages loading at mode 8767 (files|imports|types_sizes|deps|exports_file|name|compiled_files) took 46.58487ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 148.782µs 
INFO [linters_context/goanalysis] analyzers took 812.640417ms with top 10 stages: buildir: 481.157868ms, printf: 70.257765ms, nilness: 64.360526ms, typeindex: 48.605315ms, inspect: 27.956756ms, ctrlflow: 27.413866ms, inline: 27.271649ms, SA5012: 23.800341ms, fact_purity: 18.356823ms, fact_deprecated: 18.187416ms 
INFO [runner] Issues before processing: 3, after processing: 2 
INFO [runner] Processors filtering stat (in/out): cgo: 3/3, nolint_filter: 3/3, uniq_by_line: 3/2, max_from_linter: 2/2, filename_unadjuster: 3/3, exclusion_rules: 3/3, diff: 3/3, max_per_file_from_linter: 2/2, sort_results: 2/2, exclusion_paths: 3/3, fixer: 3/3, max_same_issues: 2/2, source_code: 2/2, path_shortener: 2/2, severity-rules: 2/2, path_prettifier: 2/2, path_absoluter: 3/3, invalid_issue: 3/3, path_relativity: 3/3, generated_file_filter: 3/3 
INFO [runner] processing took 152.612µs with stages: nolint_filter: 75.091µs, source_code: 38.52µs, generated_file_filter: 15.86µs, max_same_issues: 8.68µs, path_relativity: 4.36µs, uniq_by_line: 2.45µs, sort_results: 1.841µs, path_shortener: 980ns, max_from_linter: 750ns, cgo: 660ns, invalid_issue: 630ns, filename_unadjuster: 460ns, exclusion_paths: 400ns, fixer: 350ns, diff: 340ns, max_per_file_from_linter: 330ns, exclusion_rules: 330ns, path_prettifier: 260ns, path_absoluter: 230ns, severity-rules: 90ns 
INFO [runner] linters took 754.36261ms with stages: goanalysis_metalinter: 754.158458ms 
main.go:14:2: SA4023(related information): the lhs of the comparison is the 2nd return value of this function call (staticcheck)
        msg, err := example(name)
        ^
main.go:15:5: SA4023: this comparison is always true (staticcheck)
        if err != nil {
           ^
2 issues:
* staticcheck: 2
INFO File cache stats: 1 entries of total size 228B 
INFO Memory: 10 samples, avg is 167.4MB, max is 245.2MB 
INFO Execution took 804.482228ms                  
```
</details>

### A minimal reproducible example or link to a public repository

<details>

File `main.go`
```go
package main

import (
	"fmt"
	"log"
	"os"
)

func main() {
	var name string
	if len(os.Args) > 1 {
		name = os.Args[1]
	}
	msg, err := example(name)
	if err != nil {
		log.Fatalln(err)
	}
	fmt.Printf("got message %q\n", msg)
}
```

File `example_linux.go`
```go
//go:build linux

package main

import "errors"

func example(string) (string, error) {
	return "", errors.New("function not available for linux")
}
```
File `example_nonlinux.go`
```go
//go:build !linux

package main

import (
	"errors"
	"fmt"
)

func example(name string) (string, error) {
	if name == "" {
		return "", errors.New("empty names are not supported")
	}
	return fmt.Sprintf("hello %q", name), nil
}
```
</details>

### Validation

- [x] Yes, I've included all the information above (example, version, config, etc.).

### Supporter

- [ ] I am a [sponsor/backer](https://donate.golangci.org) of this project.

**Comment by ldez** [comment]
staticcheck doesn't ignore build tags, this is the opposite.

Also this is not specific to staticcheck or golangci-lint, this is how the Go tooling works.

An analysis is only viewing the element related to specific arch/os.

The analysis is comparable to a compilation.

**Comment by srebhan** [comment]
@ldez we are using golangci-lint for checking [Telegraf](https://github.com/influxdata/telegraf/pull/19541) which is available for multiple platforms and linting also happens on all the supported platforms. How can we make sure the linter passes in situations like the one in the example? You cannot use a `nolint` directive as this will fail on non-Linux platforms. I also do not see how to rewrite the code to not trigger this issue!

So the only way now is to completely disable static-check. Is this what you are saying?

**Comment by ldez** [comment]
> So the only way now is to completely disable static-check. Is this what you are saying?

No, I suggested nothing.

But I can suggest you 2 possibilities:
1. disable `nolintlint` reports:
```yml
linters:
  settings:
    nolintlint:
      allow-unused: true
```

https://golangci-lint.run/docs/linters/configuration/#nolintlint

2. use the `exclusions` configuration for this specific file and this specific report.
```yml
linters:
  exclusions:
    rules:
      - path: main.go
        text: "SA4023"
        linters:
          - staticcheck
```

https://golangci-lint.run/docs/configuration/file/#linters-configuration

**Comment by srebhan** [comment]
Thanks for your hints! I solved it using

```text
//nolint:staticcheck,nolintlint // False-positive ...
```
